### PR TITLE
feat(auth): add JWT token revocation with Redis blocklist

### LINKED ISSUE #306: Issue Template: Insecure Session Invalidation via Redis JWT Blocklist
Severity: Security / Performance

Component: apps/backend (Auth System)

Problem: Currently, the system handles user authentication via JWT tokens using @fastify/jwt. However, there is no active server-side mechanism to invalidate tokens upon logout. If a user logs out, their JWT remains cryptographically valid until its expiration time (exp). An attacker who intercepts this token can continue to make unauthorized API requests even after the legitimate user has clicked "Logout".

Proposed Solution: Implement a secure token revocation layer using the existing Redis infrastructure (redisPlugin).

Create a DELETE /auth/logout endpoint protected by the authenticate preHandler hook.

Upon logout, extract the JWT token signature from the request headers, calculate its remaining lifetime (TTL) based on its expiration timestamp, and store it in Redis with a key like blocklist:token:<signature>.

Intercept requests in the global app.decorate('authenticate') middleware hook inside app.ts to check if the incoming token exists in the Redis blocklist. If it does, throw an immediate 401 Unauthorized exception.

Please assign under gssoc 2026

### PR DESCRIPTION
**Pull request by antharya05**
## Summary

Implements secure JWT session invalidation using a Redis-backed blocklist.

Previously, JWTs remained valid until expiration even after logout, meaning a stolen token could continue accessing protected APIs after a user logged out. This change introduces server-side token revocation so tokens are immediately invalidated after logout while maintaining backward compatibility with existing logout flows.

Closes #306

---

## Type of Change

* [ ] Bug fix
* [x] New feature
* [ ] Refactor (no functional change)
* [ ] UI / Design change
* [ ] Tests only
* [ ] Documentation
* [ ] Infrastructure / DevOps
* [x] Security

---

## What Changed

* Added shared JWT utilities in `src/utils/jwt.ts`:

  * `extractRawJwt()` to consistently extract tokens from Authorization headers or cookies
  * `blocklistKey()` to generate SHA-256 hashed Redis blocklist keys

* Extended the global authentication decorator in `app.ts`:

  * Checks Redis blocklist before `jwtVerify()`
  * Rejects revoked tokens immediately
  * Registers `@fastify/cookie` before `@fastify/jwt` so cookie-based authentication works correctly for browser clients

* Added authenticated `DELETE /auth/logout` endpoint:

  * Requires a valid JWT
  * Stores a hash of the token signature in Redis
  * Uses TTL equal to the token's remaining lifetime
  * Automatically self-cleans when the token expires

* Simplified legacy `POST /auth/logout`:

  * Preserved for backward compatibility
  * Clears cookies only
  * No longer writes unverified tokens to Redis

* Added comprehensive logout test coverage:

  * Revocation flow
  * Cookie authentication
  * Redis failures
  * Edge cases
  * End-to-end revocation behavior

---

## How to Test

1. Start the backend and Redis.

```bash
npm install
npm run dev
```

2. Generate a JWT using an authenticated flow and access a protected endpoint.

3. Revoke the token:

```http
DELETE /auth/logout
Authorization: Bearer <token>
```

4. Retry the same protected request using the revoked token.

Expected result:

```http
401 Unauthorized
```

5. Run the logout test suite:

```bash
npm test -- logout
```

Expected result:

* All logout/revocation tests pass.

---

## Checklist

* [x] My code follows the project's coding style.
* [x] TypeScript compiles without errors.
* [x] I have added or updated tests for the changes I made.
* [x] All relevant tests pass locally.
* [ ] I have updated documentation where necessary.
* [x] No new console.log or debug statements left in the code.
* [x] Breaking changes are documented in this PR description.

## Additional Context

Security decisions:

* Redis blocklist entries store a SHA-256 hash of the JWT signature rather than the raw token.
* Blocklist entries expire automatically using the JWT's remaining lifetime.
* Redis failures intentionally fail open to avoid taking down authenticated requests during Redis outages; token expiration remains the fallback protection mechanism.
* `POST /auth/logout` no longer writes unverified tokens to Redis, preventing abuse of the logout endpoint.

Repository note:

While validating this feature, I compared the latest `main` branch against this branch.

Existing failures in `event.test.ts` and `team.test.ts` occur identically on `main` and were not introduced by this authentication change.

This PR introduces no additional test regressions.

### CODE REVIEW

File: apps/backend/src/routes/auth.ts [hunk 1]
```diff
@@ -286,8 +287,87 @@ export async function authRoutes(app: FastifyInstance) {
...
+        );
+      }
+    }
+
     reply.clearCookie('token', { path: '/' });
     return { message: 'Logged out' };
   });
 }
+
+function generateState(): string {
```
Comment (Harxhit) [inline apps/backend/src/routes/auth.ts]: Could you please make a util file for this.

File: apps/backend/src/routes/auth.ts [hunk 2]
```diff
@@ -286,8 +287,87 @@ export async function authRoutes(app: FastifyInstance) {
...
+  return randomBytes(32).toString('hex');
+}
+
+function buildOAuthState(clientState: string, mobileRedirectUri: string): string {
```
Comment (Harxhit) [inline apps/backend/src/routes/auth.ts]: Similarly for this to.

File: apps/backend/src/routes/auth.ts [hunk 3]
```diff
@@ -286,8 +287,87 @@ export async function authRoutes(app: FastifyInstance) {
     return { ...userData, connectedPlatforms: oauthTokens };
   });
 
-  app.post('/logout', async (request: FastifyRequest, reply: FastifyReply) => {
+  // Legacy endpoint kept for backward compatibility with existing clients.
+  // Cookie-only logout — use DELETE /auth/logout for token revocation.
+  app.post('/logout', async (_request: FastifyRequest, reply: FastifyReply) => {
+    reply.clearCookie('token', { path: '/' });
+    return { message: 'Logged out' };
```
Comment (Harxhit) [inline apps/backend/src/routes/auth.ts:294]: Could you please add logging here.

File: apps/backend/src/routes/auth.ts [hunk 4]
```diff
@@ -286,8 +287,87 @@ export async function authRoutes(app: FastifyInstance) {
...
+    return generateState();
+  }
+  if (clientState.startsWith('mobile_') && mobileRedirectUri) {
+    const encodedRedirect = Buffer.from(mobileRedirectUri, 'utf8').toString('base64url');
+    return `${clientState}.${encodedRedirect}.${generateState()}`;
+  }
+  return `${clientState}.${generateState()}`;
+}
+
+function getMobileRedirectUri(state?: string): string | null {
```
Comment (Harxhit) [inline apps/backend/src/routes/auth.ts]: Similarly for this too please.

File: apps/backend/src/routes/auth.ts
Comment (Harxhit) [inline apps/backend/src/routes/auth.ts:1]: Should we include ``userId `` in the warning log to make Redis token revocation failures easier to trace?

**Review by Harxhit — APPROVED** [review]
LGTM. All review comments have been addressed, and I verified that linting, type checking, and unit tests are passing locally.

Approving.
